<img style="float: left; margin: 30px 15px 15px 15px;" src="https://encrypted-tbn0.gstatic.com/images?q=tbn:ANd9GcTFzQj91sOlkeDFkg5HDbjtR4QJYmLXkfMNig&usqp=CAU" width="400" height="500" /> 
    
    
### <font color='navy'> Simulación matemática. 

**Nombres:** Santiago Reyes, Juan Pablo Barba, Emilio Andrade

**Fecha:** 1 de noviembre, 2023

**Expediente** : 745826, 747297.
    
**Profesor:** Oscar David Jaramillo Zuluaga.
    
**Link Github**: Link con el enlace del repositorio del creador del proyecto

# Tarea: 8 Clase: 18 

## Tarea.
> - Consultar tasas de interés reales en algún banco y proyectar un ahorro mensual para que al terminar su carrera tengan $\$85000$ MXN en su cuenta.
> - Hacer esto en un nuevo archivo, llamarlo Gi_T7ApellidoNombre.ipynb y subirlo a canvas. Donde $i=1,2$ es el número del grupo al que pertenece 
> - Plazo: 
> # Entrega 14 de noviembre a las 19:59 en parejas

Código Solución Estudiante 1

In [2]:
# Librería de análisis de datos
import pandas as pd
# Librería para manipulación de fechas
from datetime import date
import numpy as np
import numpy_financial as npf

In [3]:
# Datos para la cuenta de ahorro
tasa_interes = 0.0401
years = 3.5
Abono_year = 12
inicial = 0
meta = 85000
fecha_inicio = (date(2024,1,1))

In [4]:
# Cálculo de abonos mensuales iguales para llegar a la meta de ahorro en el tiempo deseado
Abono = -npf.pmt(tasa_interes/Abono_year, years*Abono_year, (meta-inicial))
Abono

2172.525406361609

In [5]:
# Rango de fechas en la tabla
rng = pd.date_range(fecha_inicio, periods=years*Abono_year, freq='M')
rng.name = 'Fecha del abono'
# Columnas de abonos y extracto de cuenta en las fechas respectivas
df = pd.DataFrame(index = rng, columns=['Abono','Balance'],dtype='float')
# Para poner el indice con los periodos y no con las fechas
df.reset_index(inplace=True)

df.index+=1
df.index.name='Periodo'
df.head()

Fecha del abono  Abono  Balance
Periodo                                
1            2024-01-31    NaN      NaN
2            2024-02-29    NaN      NaN
3            2024-03-31    NaN      NaN
4            2024-04-30    NaN      NaN
5            2024-05-31    NaN      NaN

In [6]:
# Los abonos son iguales
df['Abono'] = Abono
# Vector indizador de los periodos de los abonos
index_vector = np.arange(1,len(rng)+1)
# Extracto de la cuenta mes a mes
df["Balance"] = inicial * (1 + tasa_interes/Abono_year)**index_vector + \
                Abono * (((1 + tasa_interes/Abono_year)**index_vector - 1)/(tasa_interes/Abono_year))

df.head()

Fecha del abono        Abono       Balance
Periodo                                           
1            2024-01-31  2172.525406   2172.525406
2            2024-02-29  2172.525406   4352.310668
3            2024-03-31  2172.525406   6539.380046
4            2024-04-30  2172.525406   8733.757881
5            2024-05-31  2172.525406  10935.468595

In [7]:
# Redondear con dos cifras decimales y mostrar DataFrame
df = df.round(2)
df

Fecha del abono    Abono   Balance
Periodo                                   
1            2024-01-31  2172.53   2172.53
2            2024-02-29  2172.53   4352.31
3            2024-03-31  2172.53   6539.38
4            2024-04-30  2172.53   8733.76
5            2024-05-31  2172.53  10935.47
6            2024-06-30  2172.53  13144.54
7            2024-07-31  2172.53  15360.99
8            2024-08-31  2172.53  17584.84
9            2024-09-30  2172.53  19816.13
10           2024-10-31  2172.53  22054.88
11           2024-11-30  2172.53  24301.10
12           2024-12-31  2172.53  26554.83
13           2025-01-31  2172.53  28816.10
14           2025-02-28  2172.53  31084.91
15           2025-03-31  2172.53  33361.32
16           2025-04-30  2172.53  35645.32
17           2025-05-31  2172.53  37936.96
18           2025-06-30  2172.53  40236.26
19           2025-07-31  2172.53  42543.24
20           2025-08-31  2172.53  44857.93
21           2025-09-30  2172.53  47180.36
22           2025-10-31  2172.53  49510.55
23           2025-11-30  2172.53  51848.52
24           2025-12-31  2172.53  54194.31
25           2026-01-31  2172.53  56547.93
26           2026-02-28  2172.53  58909.42
27           2026-03-31  2172.53  61278.80
28           2026-04-30  2172.53  63656.10
29           2026-05-31  2172.53  66041.34
30           2026-06-30  2172.53  68434.56
31           2026-07-31  2172.53  70835.77
32           2026-08-31  2172.53  73245.00
33           2026-09-30  2172.53  75662.29
34           2026-10-31  2172.53  78087.65
35           2026-11-30  2172.53  80521.12
36           2026-12-31  2172.53  82962.72
37           2027-01-31  2172.53  85412.48
38           2027-02-28  2172.53  87870.42
39           2027-03-31  2172.53  90336.58
40           2027-04-30  2172.53  92810.98
41           2027-05-31  2172.53  95293.65
42           2027-06-30  2172.53  97784.62

Código Solución Estudiante 2

In [15]:
#Importar librerías
import pandas as pd
from datetime import date
import numpy as np
import numpy_financial as npf

In [16]:
# Datos para la cuenta de ahorro
tasa_interes = 0.05  # 5% tasa de interés anual
years = 5  # Duración del plan de ahorro en años
Abono_year = 12  # Abonos por año (mensuales)
inicial = 0  # Balance inicial de la cuenta
meta = 85000  # Meta de ahorro en MXN
fecha_inicio = date(2023, 11, 1)  # Fecha de inicio del plan de ahorro

In [17]:
# Cálculo de abonos mensuales iguales para llegar a la meta de ahorro en el tiempo deseado
Abono = -npf.pmt(tasa_interes/Abono_year, years*Abono_year, (meta-inicial))
Abono

1604.0548597409343

In [18]:
# Rango de fechas en la tabla
rng = pd.date_range(fecha_inicio, periods=years*Abono_year, freq='M')
rng.name = 'Fecha del abono'

# Columnas de abonos y extracto de cuenta en las fechas respectivas
df = pd.DataFrame(index=rng, columns=['Abono', 'Balance'], dtype='float')

# Para poner el índice con los periodos
df.reset_index(inplace=True)
df.index += 1
df.index.name = 'Periodo'

In [19]:
# Ponemos que en abono siempre sea la misma cantidad
df['Abono'] = Abono

# Vector indizador de los periodos de los abonos
index_vector = np.arange(1, len(rng) + 1)

# Suma de la cuenta mes a mes
df["Balance"] = inicial * (1 + tasa_interes/Abono_year)**index_vector + \
                Abono * (((1 + tasa_interes/Abono_year)**index_vector - 1) / (tasa_interes/Abono_year))
df.head()

Fecha del abono       Abono      Balance
Periodo                                         
1            2023-11-30  1604.05486  1604.054860
2            2023-12-31  1604.05486  3214.793281
3            2024-01-31  1604.05486  4832.243113
4            2024-02-29  1604.05486  6456.432319
5            2024-03-31  1604.05486  8087.388980

In [20]:
# Redondear con dos cifras decimales y mostrar DataFrame
df = df.round(2)

df

Fecha del abono    Abono    Balance
Periodo                                    
1            2023-11-30  1604.05    1604.05
2            2023-12-31  1604.05    3214.79
3            2024-01-31  1604.05    4832.24
4            2024-02-29  1604.05    6456.43
5            2024-03-31  1604.05    8087.39
6            2024-04-30  1604.05    9725.14
7            2024-05-31  1604.05   11369.72
8            2024-06-30  1604.05   13021.15
9            2024-07-31  1604.05   14679.46
10           2024-08-31  1604.05   16344.68
11           2024-09-30  1604.05   18016.83
12           2024-10-31  1604.05   19695.96
13           2024-11-30  1604.05   21382.08
14           2024-12-31  1604.05   23075.23
15           2025-01-31  1604.05   24775.43
16           2025-02-28  1604.05   26482.71
17           2025-03-31  1604.05   28197.11
18           2025-04-30  1604.05   29918.66
19           2025-05-31  1604.05   31647.37
20           2025-06-30  1604.05   33383.29
21           2025-07-31  1604.05   35126.44
22           2025-08-31  1604.05   36876.86
23           2025-09-30  1604.05   38634.57
24           2025-10-31  1604.05   40399.60
25           2025-11-30  1604.05   42171.98
26           2025-12-31  1604.05   43951.76
27           2026-01-31  1604.05   45738.94
28           2026-02-28  1604.05   47533.58
29           2026-03-31  1604.05   49335.69
30           2026-04-30  1604.05   51145.31
31           2026-05-31  1604.05   52962.47
32           2026-06-30  1604.05   54787.20
33           2026-07-31  1604.05   56619.54
34           2026-08-31  1604.05   58459.51
35           2026-09-30  1604.05   60307.14
36           2026-10-31  1604.05   62162.48
37           2026-11-30  1604.05   64025.54
38           2026-12-31  1604.05   65896.37
39           2027-01-31  1604.05   67774.99
40           2027-02-28  1604.05   69661.44
41           2027-03-31  1604.05   71555.75
42           2027-04-30  1604.05   73457.96
43           2027-05-31  1604.05   75368.09
44           2027-06-30  1604.05   77286.18
45           2027-07-31  1604.05   79212.26
46           2027-08-31  1604.05   81146.36
47           2027-09-30  1604.05   83088.53
48           2027-10-31  1604.05   85038.78
49           2027-11-30  1604.05   86997.17
50           2027-12-31  1604.05   88963.71
51           2028-01-31  1604.05   90938.45
52           2028-02-29  1604.05   92921.41
53           2028-03-31  1604.05   94912.64
54           2028-04-30  1604.05   96912.16
55           2028-05-31  1604.05   98920.02
56           2028-06-30  1604.05  100936.24
57           2028-07-31  1604.05  102960.86
58           2028-08-31  1604.05  104993.92
59           2028-09-30  1604.05  107035.45
60           2028-10-31  1604.05  109085.49

#### Conclusión
Con base en la tabla generada, puedes observar que necesitarías hacer abonos mensuales de $\$1,604.05$ MXN para alcanzar tu meta de ahorro de $\$85,000$ MXN en un plazo de 5 años con una tasa de interés del 5% anual.

La columna "Balance" muestra cómo se acumulará tu ahorro con el tiempo, teniendo en cuenta los abonos mensuales y la tasa de interés. Al final del periodo (Periodo 60, que corresponde a octubre de 2028), habrás alcanzado tu meta de $\$109,085.49 MXN$, superando incluso los $\$85,000$ MXN iniciales gracias al efecto del interés compuesto.

Es importante recalcar que estos cálculos son una estimación y pueden variar en función de cambios en la tasa de interés, entre otros factores. Sin embargo, te proporcionan un buen punto de partida para tu plan de ahorro.

Código Solución Estudiante 3